tests/test_typechecker.py

```diff
@@ -4652,5 +4652,63 @@ def test_inner_class(self):
         self.assertRaises(InputTypeError, lambda:
                 py3.test_inner_class_testf1_err())
 
+
+class Test_utils(unittest.TestCase):
+    def test_empty_values(self):
+        self.assertTrue(pytypes.is_of_type([], typing.Sequence))
+        self.assertTrue(pytypes.is_of_type([], typing.Sequence[int]))
+
+        self.assertTrue(isinstance(set(), typing.Sized))
+        self.assertTrue(pytypes.is_of_type(set(), typing.Sized))
+        self.assertTrue(isinstance([], typing.Sized))
+        self.assertTrue(pytypes.is_of_type([], typing.Sized))
+
+    def test_tuple_elipsis(self):
+        class Foo:
+            pass
+
+        self.assertTrue(pytypes.is_subtype(typing.Tuple[Foo], typing.Tuple[object, ...]))
+        self.assertTrue(pytypes.is_subtype(typing.Tuple[Foo], typing.Tuple[typing.Any, ...]))
+
+    def test_bound_typevars_readonly(self):
+        T = typing.TypeVar('T', covariant=True)
+
+        class L(typing.List[T]):
+            pass
+
+        C = typing.TypeVar('T', bound=L)
+
+        self.assertTrue(pytypes.is_subtype(L[float], C))
+        self.assertTrue(pytypes.is_subtype(L[float], C, bound_typevars={}))
+        self.assertFalse(pytypes.is_subtype(L[float], C, bound_typevars_readonly=True, bound_typevars={}))
+        self.assertTrue(pytypes.is_subtype(L[float], C, bound_typevars_readonly=False, bound_typevars={}))
+
+    def test_forward_declaration(self):
+        Container = typing.Union[
+            typing.List['Data'],
+        ]
+
+        Data = typing.Union[
+            Container,
+            str, bytes, bool, float, int, dict,
+        ]
+
+        with self.assertRaises(pytypes.ForwardRefError):
+            pytypes.is_subtype(typing.List[float], Container)
+
+        pytypes.resolve_fw_decl(Container)
+
+        pytypes.is_subtype(typing.List[float], Container)
+
+    def test_forward_declaration_infinite_recursion(self):
+        Data = typing.Union['Container', float]
+        Container = typing.Union[Data, int]
+
+        pytypes.resolve_fw_decl(Data)
+        pytypes.resolve_fw_decl(Container)
+
+        self.assertFalse(pytypes.is_subtype(list, Container))
```

Comment:
I think this should raise an exception.

Reply:
I'd like to pretend that this succeeds because of recursion proof handling in ``type_util._issubclass_Union``/``type_util._issubclass_Union_rec``. However I tried to out-comment the recursion-proof part and still get the same result. While this behavior is puzzling I would not investigate further unless we find a particular failure.

Reply:
So you are saying you disable recursion checks you added and this test still succeeds?

I copied it from here and at that time it failed for sure: https://github.com/Stewori/pytypes/issues/22#issuecomment-354412818

Reply:
Oh right, the exception is what you actually suggested. Do you find it important to uncover recursive type declarations this way?
I think it would be possible to let ``type_util._issubclass_Union`` throw an exception in such a case. I think I'd prefer to have such a behavior optional. Will think about it...

Reply:
I do not have strong opinion here. I just see it as an invalid type. So it would simplify debugging. But maybe this type is not invalid and it is true that isinstance should return False, if we would be able to resolve the recursion somehow. So I do not really mind either way. It just be that it will surprise some.